ext/curl/share.c

```diff
@@ -134,6 +136,159 @@ PHP_FUNCTION(curl_share_strerror)
 }
 /* }}} */
 
+/**
+ * Initialize a persistent curl share handle with the given share options, reusing an existing one if found.
+ *
+ * Throws an exception if the share options are invalid.
+ */
+PHP_FUNCTION(curl_share_init_persistent)
+{
+	zval *share_opts = NULL, *entry = NULL;
+	zend_ulong persistent_id = 0;
+
+	php_curlsh *sh = NULL;
+
+	CURLSHcode error = 0;
+
+	ZEND_PARSE_PARAMETERS_START(1, 1)
+		Z_PARAM_ARRAY(share_opts)
+	ZEND_PARSE_PARAMETERS_END();
+
+	ZEND_HASH_FOREACH_VAL(Z_ARRVAL_P(share_opts), entry) {
+		ZVAL_DEREF(entry);
+
+		bool       failed = false;
+		zend_ulong option = zval_try_get_long(entry, &failed);
+
+		if (failed) {
+			zend_argument_type_error(1, "must contain only integer values, %s given", zend_zval_value_name(entry));
+			goto error;
+		}
+
+		switch (option) {
+			// Disallowed options
+			case CURL_LOCK_DATA_COOKIE:
+				zend_argument_value_error(1, "must not contain CURL_LOCK_DATA_COOKIE because it is unsafe");
+				goto error;
+
+			// Allowed options
+			case CURL_LOCK_DATA_DNS:
+				persistent_id |= 1 << 0;
+				break;
+			case CURL_LOCK_DATA_SSL_SESSION:
+				persistent_id |= 1 << 1;
+				break;
+			case CURL_LOCK_DATA_CONNECT:
+				persistent_id |= 1 << 2;
+				break;
+			case CURL_LOCK_DATA_PSL:
+				persistent_id |= 1 << 3;
+				break;
+
+			// Unknown options
+			default:
+				zend_argument_value_error(1, "must contain only CURL_LOCK_DATA_* constants");
+				goto error;
+		}
+	} ZEND_HASH_FOREACH_END();
+
+	// Construct a property field for the CurlSharePersistentHandle object with the enabled share options.
+	zval share_opts_prop;
+	array_init(&share_opts_prop);
+
+	if (persistent_id & 1 << 0) {
+		add_next_index_long(&share_opts_prop, CURL_LOCK_DATA_DNS);
+	}
+
+	if (persistent_id & 1 << 1) {
+		add_next_index_long(&share_opts_prop, CURL_LOCK_DATA_SSL_SESSION);
+	}
+
+	if (persistent_id & 1 << 2) {
+		add_next_index_long(&share_opts_prop, CURL_LOCK_DATA_CONNECT);
+	}
+
+	if (persistent_id & 1 << 3) {
+		add_next_index_long(&share_opts_prop, CURL_LOCK_DATA_PSL);
+	}
+
+	object_init_ex(return_value, curl_share_persistent_ce);
+
+	zend_update_property(
+	    curl_share_persistent_ce,
+		Z_OBJ_P(return_value),
+		"options",
+		sizeof("options") - 1,
```

Comment:
```suggestion
		strlen("options"),
```
`strlen()` in new code.

Reply:
Or even better: ZEND_STRL("options")

Reply:
Will do - though while we're talking about this, is this the correct way to update a PHP object's properties in C? I mean no offense to the C codebase, but it feels a little error-prone to have to pass in a string and specify the string's length, instead of maybe referring to the `options` property statically.

Reply:
> is this the correct way to update a PHP object's properties in C?

AFAICT that's the correct way that handles all the edge cases (e.g. properly checking for repeated `readonly` assignments), yes. You could technically also access the property directly via `OBJ_PROP_NUM`, but that is really not any safer (on the contrary, probably):

https://github.com/php/php-src/blob/b7c76ded31cbb8431f03555b99f2e2e600a43109/Zend/zend_attributes.c#L106C12-L106C24